java-manta-client/src/main/java/com/joyent/manta/client/MantaClient.java

```diff
@@ -1493,46 +1499,80 @@ public void move(final String source, final String destination)
         MantaObjectResponse entry = head(source);
 
         if (entry.isDirectory()) {
-            putDirectory(destination, true);
-            MantaHttpHeaders sourceHeaders = entry.getHttpHeaders();
-            Long contentsCount = sourceHeaders.getResultSetSize();
-
-            // If we were just copying an empty directory, we just create the
-            // new directory and delete the original
-            if (contentsCount != null && contentsCount == 0L) {
-                delete(source);
-                return;
-            }
+            moveDirectory(source, destination, entry);
+        } else {
+            moveFile(source, destination);
+        }
+    }
 
-            MantaObjectResponse destDir = head(destination);
-            String destDirPath = destDir.getPath();
-            String sourceDirPath = entry.getPath();
+    /**
+     * Moves a file from one path to another path. When moving
+     * files between different directories, this operation is
+     * not transactional and will fail or produce inconsistent result if
+     * the source or the destination is modified while the operation is
+     * in progress.
+     *
+     * @param source Original path to move from
+     * @param destination Destination path to move to
+     *
+     * @throws IOException thrown when something goes wrong
+     */
+    private void moveFile(final String source, final String destination) throws IOException {
+        final String destinationDir = FilenameUtils.getFullPath(destination);
```

Comment:
Is this going to do the right thing on a path where `formatPath` has not been called yet?